src/Stats.AzureCdnLogs.Common/AzureHelpers/AzureBlobLeaseManager.cs

```
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+
+namespace Stats.AzureCdnLogs.Common
+{
+    public class AzureBlobLeaseManager
+    {
+        private ConcurrentDictionary<Uri, string> _leasedBlobs = new ConcurrentDictionary<Uri, string>();
+        private BlobRequestOptions _blobRequestOptions;
+
+        public AzureBlobLeaseManager(BlobRequestOptions blobRequestOptions=null)
+        {
+            _blobRequestOptions = blobRequestOptions;
+        }
+
+        /// <summary>
+        /// Try to aquire a lease on the blob. If the aquire is sucessful the lease will be renewed at every 60 seconds. 
+        /// In order to stop the renew task the <see cref="Stats.AzureCdnLogs.Common.AzureBlobLeaseManager.TryReleaseLease(CloudBlob)"/> needs to be invoked
+        /// or the token to be cancelled.
+        /// </summary>
+        /// <param name="blob">The blob to aquire the lease on.</param>
+        /// <param name="token">A token to cancel the operation.</param>
+        /// <param name="renewStatusTask">The renew task.</param>
+        /// <returns></returns>
+        public bool AcquireLease(CloudBlob blob, CancellationToken token, out Task<bool> renewStatusTask)
+        {
+            int maxRenewPeriodInSeconds = 60;
+            renewStatusTask = null;
+            blob.FetchAttributes();
+            if (token.IsCancellationRequested || blob.Properties.LeaseStatus == LeaseStatus.Locked)
+            {
+                return false;
+            }
+            var proposedLeaseId = Guid.NewGuid().ToString();
+            string leaseId;
+
+            leaseId = blob.AcquireLease(TimeSpan.FromSeconds(maxRenewPeriodInSeconds), proposedLeaseId);
```

>AcquireLease [](start = 27, length = 12)

It seems that `AcquireLease` can throw a `StorageException` exception. Should you catch and return false?

---

The model that I follow is that i let the exemptions to bubble up and I catch them at a  upper level from where I log them. Otherwise I will catch all exceptions at this level, return false in case of exception and rename the method Try*** (in this case TryAcquireLease).

---

Makes sense. What about exceptions thrown from the renewal threads though? How are those handled?

---

They are not at this level. The caller will need to wait/handle any exceptions as needed. I added code to log the exceptions on these tasks.